### PR TITLE
Improve kernel launch performance

### PR DESCRIPTION
**Pull request by okuta**
Creating Python object is high cost. This PR reduce it.
- Use convert_scalar to simplify.
- Use PyObject_Malloc to avoid MAX_NDIM limit.

### CODE REVIEW

**Comment by niboshi** [comment]
Could you answer some questions:
- Why did you refactor class inheritance? It does not seem to be directly related to Python object reduction?
- Why did you replace an array with a vector? I think vector has more overhead.

**Comment by okuta** [comment]
Thank you for your review.
I will split this PR.

**Comment by okuta** [comment]
I removed `vector`.

**Comment by okuta** [comment]
This PR reduce a little processing time. I have to use more aggressive way.
I will improve this.

File: cupy/cuda/function.pyx [hunk 1]
```diff
@@ -119,12 +34,16 @@ cdef _launch(size_t func, Py_ssize_t grid0, int grid1, int grid2,
...
-    cdef CPointer cp
-    kargs.reserve(len(args))
-    for a in args:
+    cdef core.CPointer cp
+    cdef int i
+
+    kargs.resize(len(args))
+    for i, a in enumerate(args):
+        if a is None:
+            continue
```
Comment (niboshi) [inline cupy/cuda/function.pyx]: How about `kargs.push_back(nullptr)` here, instead of `kargs.resize()`? I think it is easier to understand.

File: cupy/core/carray.pxi [hunk 2]
```diff
@@ -1,67 +1,60 @@
...
-    cdef:
-        _CArray val
-
     def __init__(self, ndarray arr):
-        cdef Py_ssize_t i
-        cdef int ndim = arr._shape.size()
-        self.val.data = <void*>arr.data.ptr
-        self.val.size = arr.size
+        cdef int i, ndim = arr._shape.size()
+        self.ptr = cpython.PyObject_Malloc(sizeof(Py_ssize_t) * (2 + 2 * ndim))
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/carray.pxi]: https://docs.python.org/3.6/c-api/memory.html#the-pymalloc-allocator
> Python has a pymalloc allocator optimized for small objects (smaller or equal to 512 bytes) with a short lifetime. 
Current CArray and CIndexter is smaller than 512 bytes. OK, I will revert,

File: cupy/cuda/function.pyx (hunk 1 above)
Comment (okuta) [reply cupy/cuda/function.pyx]: OK

File: cupy/core/carray.pxi (hunk 2 above)
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/core/carray.pxi]: It's ok if it doesn't degrade performance, but even so, we need some comments.

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/carray.pxi [hunk 3]
```diff
@@ -30,38 +22,23 @@ cdef class CArray(CPointer):
...
+cdef class Indexer(CPointer):
 
-
-cdef class Indexer:
     def __init__(self, tuple shape):
-        cdef Py_ssize_t size = 1
-        for s in shape:
-            size *= s
+        cdef Py_ssize_t s, size = 1
+        cdef int i, ndim = len(shape)
```
Comment (niboshi) [inline cupy/core/carray.pxi]: `ndim` is unused.

**Comment by niboshi** [comment]
Please check test failure.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins test this please.
@niboshi Cloud you merge this PR?

File: cupy/core/carray.pxi [hunk 4]
```diff
@@ -1,12 +1,11 @@
+cimport cpython  # NOQA
```
Comment (niboshi) [inline cupy/core/carray.pxi]: Seems like this import is not needed.

File: cupy/cuda/function.pyx [hunk 5]
```diff
@@ -122,6 +43,8 @@ cdef _launch(size_t func, Py_ssize_t grid0, int grid1, int grid2,
     cdef CPointer cp
     kargs.reserve(len(args))
     for a in args:
+        if a is None:
+            kargs.push_back(<void*>0)
```
Comment (niboshi) [inline cupy/cuda/function.pyx]: Is this relevant to this PR?

Comment (okuta) [reply cupy/cuda/function.pyx]: Yes.

File: cupy/core/carray.pxi (hunk 4 above)
Comment (okuta) [reply cupy/core/carray.pxi]: Thanks.

File: cupy/cuda/function.pyx (hunk 5 above)
Comment (niboshi) [reply cupy/cuda/function.pyx]: Could you explain how?

Could you also answer this question?
>Is the title "Reduce Python object to improve kernel launch performance" still relevant?
>Could you explain where Python objects are reduced, or fix the title accordingly?

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/cuda/function.pyx (hunk 5 above)
Comment (okuta) [reply cupy/cuda/function.pyx]: OK, I changed PR title.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Comment by kmaehashi** [comment]
This PR has been reverted in #2185
